pkg/util/taints/taints_test.go

```diff
@@ -25,39 +25,76 @@ import (
 )
 
 func TestAddOrUpdateTaint(t *testing.T) {
-	node := &v1.Node{}
-
-	taint := &v1.Taint{
+	taint := v1.Taint{
 		Key:    "foo",
 		Value:  "bar",
 		Effect: v1.TaintEffectNoSchedule,
 	}
 
-	checkResult := func(testCaseName string, newNode *v1.Node, expectedTaint *v1.Taint, result, expectedResult bool, err error) {
+	taintNew := v1.Taint{
+		Key:    "foo_1",
+		Value:  "bar_1",
+		Effect: v1.TaintEffectNoSchedule,
+	}
+
+	taintUpdateValue := taint
+	taintUpdateValue.Value = "bar_1"
+
+	testcases := []struct {
+		name           string
+		node           *v1.Node
+		taint          *v1.Taint
+		expectedUpdate bool
+		expectedTaints []v1.Taint
+	}{
+		{
+			name:           "add a new taint",
+			node:           &v1.Node{},
+			taint:          &taint,
+			expectedUpdate: true,
+			expectedTaints: []v1.Taint{taint},
+		},
+		{
+			name: "add a unique taint",
+			node: &v1.Node{
+				Spec: v1.NodeSpec{Taints: []v1.Taint{taint}},
+			},
+			taint:          &taintNew,
+			expectedUpdate: true,
+			expectedTaints: []v1.Taint{taint, taintNew},
+		},
+		{
+			name: "add duplicate taint",
+			node: &v1.Node{
+				Spec: v1.NodeSpec{Taints: []v1.Taint{taint}},
+			},
+			taint:          &taint,
+			expectedUpdate: false,
+			expectedTaints: []v1.Taint{taint},
+		},
+		{
+			name: "update taint value",
+			node: &v1.Node{
+				Spec: v1.NodeSpec{Taints: []v1.Taint{taint}},
+			},
+			taint:          &taintUpdateValue,
+			expectedUpdate: true,
+			expectedTaints: []v1.Taint{taintUpdateValue},
+		},
+	}
+
+	for _, tc := range testcases {
+		newNode, updated, err := AddOrUpdateTaint(tc.node, tc.taint)
 		if err != nil {
-			t.Errorf("[%s] should not raise error but got %v", testCaseName, err)
+			t.Errorf("[%s] should not raise error but got %v", tc.name, err)
 		}
-		if result != expectedResult {
-			t.Errorf("[%s] should return %t, but got: %t", testCaseName, expectedResult, result)
+		if updated != tc.expectedUpdate {
+			t.Errorf("[%s] expected taints to not be updated", tc.name)
 		}
-		if len(newNode.Spec.Taints) != 1 || !reflect.DeepEqual(newNode.Spec.Taints[0], *expectedTaint) {
-			t.Errorf("[%s] node should only have one taint: %v, but got: %v", testCaseName, *expectedTaint, newNode.Spec.Taints)
+		if len(newNode.Spec.Taints) != len(tc.expectedTaints) || !reflect.DeepEqual(newNode.Spec.Taints, tc.expectedTaints) {
```

Comment:
Use cmp.Diff

Reply:
The `reflect.DeepEqual` function is used in this test package for comparison. If it needs to be changed, does it need to be modified uniformly?

related: https://github.com/kubernetes/kubernetes/pull/108332#discussion_r903448794

Reply:
you can replace previous usages in a follow up PR

Reply:
OK, got it.